services/sqs/src/main/java/software/amazon/awssdk/services/sqs/internal/batchmanager/QueueAttributesManager.java

```diff
@@ -0,0 +1,157 @@
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.sqs.internal.batchmanager;
+
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.services.sqs.SqsAsyncClient;
+import software.amazon.awssdk.services.sqs.model.GetQueueAttributesRequest;
+import software.amazon.awssdk.services.sqs.model.QueueAttributeName;
+import software.amazon.awssdk.services.sqs.model.ReceiveMessageRequest;
+import software.amazon.awssdk.utils.Validate;
+
+@SdkInternalApi
+public final class QueueAttributesManager {
+
+    private static final List<QueueAttributeName> QUEUE_ATTRIBUTE_NAMES =
+        Arrays.asList(QueueAttributeName.RECEIVE_MESSAGE_WAIT_TIME_SECONDS,
+                      QueueAttributeName.VISIBILITY_TIMEOUT);
+    private final SqsAsyncClient sqsClient;
+    private final String queueUrl;
+    private final AtomicReference<CompletableFuture<Map<QueueAttributeName, String>>> queueAttributeMap = new AtomicReference<>();
+
+    public QueueAttributesManager(SqsAsyncClient sqsClient, String queueUrl) {
+        this.sqsClient = sqsClient;
+        this.queueUrl = queueUrl;
+    }
+
+    /**
+     * Retrieves the received message timeout based on the provided request and queue attributes.
+     *
+     * @param rq                 The receive message request
+     * @param configuredWaitTime The configured minimum wait time
+     * @return CompletableFuture with the calculated receive message timeout in milliseconds
+     */
+    public CompletableFuture<Duration> getReceiveMessageTimeout(ReceiveMessageRequest rq, Duration configuredWaitTime) {
+        Integer waitTimeSeconds = rq.waitTimeSeconds();
+        if (waitTimeSeconds != null) {
+            long waitTimeMillis = TimeUnit.SECONDS.toMillis(waitTimeSeconds);
+            return CompletableFuture.completedFuture(Duration.ofMillis(Math.max(configuredWaitTime.toMillis(), waitTimeMillis)));
+        }
+
+        CompletableFuture<Map<QueueAttributeName, String>> attributeFuture = getAttributeMap();
+        CompletableFuture<Duration> resultFuture = attributeFuture.thenApply(attributes -> {
+            String waitTimeSecondsStr = attributes.get(QueueAttributeName.RECEIVE_MESSAGE_WAIT_TIME_SECONDS);
+            long waitTimeFromSqsMillis = TimeUnit.SECONDS.toMillis(Long.parseLong(waitTimeSecondsStr));
+            return Duration.ofMillis(Math.max(configuredWaitTime.toMillis(), waitTimeFromSqsMillis));
+        });
+
+        resultFuture.whenComplete((r, t) -> {
+            if (resultFuture.isCancelled()) {
```

Comment:
minor: probably better to just check `instanceof` on `t`. If this future was cancelled, `t` will have type `CancellationException`.